File: services/whatsapp.py
import os
import requests

_BASE = f"https://graph.facebook.com/{os.environ.get('WHATSAPP_API_VERSION', 'v20.0')}"
_PHONE_ID = os.environ["WHATSAPP_PHONE_NUMBER_ID"]
_TOKEN = os.environ["WHATSAPP_TOKEN"]
_HEADERS = {"Authorization": f"Bearer {_TOKEN}", "Content-Type": "application/json"}


def send_message(to: str, text: str) -> None:
    requests.post(
        f"{_BASE}/{_PHONE_ID}/messages",
        json={
            "messaging_product": "whatsapp",
            "to": to,
            "type": "text",
            "text": {"body": text},
        },
        headers=_HEADERS,
        timeout=10,
    ).raise_for_status()


def send_button_message(to: str, body_text: str, buttons: list[str]) -> None:
    """Envía mensaje con hasta 3 botones de respuesta rápida.
    Hace fallback a texto plano si la API rechaza el interactivo.
    """
    try:
        requests.post(
            f"{_BASE}/{_PHONE_ID}/messages",
            json={
                "messaging_product": "whatsapp",
                "to": to,
                "type": "interactive",
                "interactive": {
                    "type": "button",
                    "body": {"text": body_text},
                    "action": {
                        "buttons": [
                            {"type": "reply", "reply": {"id": f"btn_{i}", "title": label[:20]}}
                            for i, label in enumerate(buttons[:3])
                        ]
                    },
                },
            },
            headers=_HEADERS,
            timeout=10,
        ).raise_for_status()
    except Exception:
        # Fallback: texto plano con opciones
        send_message(to, f"{body_text}\n\nOpciones: {' / '.join(buttons)}")


def get_media_url(media_id: str) -> str:
    resp = requests.get(
        f"{_BASE}/{media_id}",
        headers={"Authorization": f"Bearer {_TOKEN}"},
        timeout=10,
    )
    resp.raise_for_status()
    return resp.json()["url"]


def download_media(media_url: str) -> bytes:
    resp = requests.get(
        media_url,
        headers={"Authorization": f"Bearer {_TOKEN}"},
        timeout=30,
    )
    resp.raise_for_status()
    return resp.content


def mark_as_read(message_id: str) -> None:
    """Marca el mensaje como leído (doble check azul). No crítico."""
    try:
        requests.post(
            f"{_BASE}/{_PHONE_ID}/messages",
            json={
                "messaging_product": "whatsapp",
                "status": "read",
                "message_id": message_id,
            },
            headers=_HEADERS,
            timeout=5,
        )
    except Exception:
        pass

File: src/services/whatsapp.js
'use strict';

const axios = require('axios');
const config = require('../config');

const BASE_URL = `https://graph.facebook.com/${config.whatsapp.version}`;

/** Headers comunes para todas las llamadas a la API de Meta */
function authHeaders() {
  return {
    Authorization: `Bearer ${config.whatsapp.token}`,
    'Content-Type': 'application/json',
  };
}

/**
 * Envía un mensaje de texto simple.
 * @param {string} to  Número en formato E.164 (ej. 5491112345678)
 * @param {string} text
 */
async function sendMessage(to, text) {
  await axios.post(
    `${BASE_URL}/${config.whatsapp.phoneNumberId}/messages`,
    {
      messaging_product: 'whatsapp',
      to,
      type: 'text',
      text: { body: text },
    },
    { headers: authHeaders() }
  );
}

/**
 * Envía un mensaje con botones de respuesta rápida (máx 3 botones).
 * Fallback a texto si la plataforma no soporta interactivos.
 * @param {string} to
 * @param {string} bodyText   Cuerpo del mensaje
 * @param {string[]} buttons  Etiquetas de los botones (máx 20 chars c/u)
 */
async function sendButtonMessage(to, bodyText, buttons) {
  try {
    await axios.post(
      `${BASE_URL}/${config.whatsapp.phoneNumberId}/messages`,
      {
        messaging_product: 'whatsapp',
        to,
        type: 'interactive',
        interactive: {
          type: 'button',
          body: { text: bodyText },
          action: {
            buttons: buttons.slice(0, 3).map((label, i) => ({
              type: 'reply',
              reply: { id: `btn_${i}_${Date.now()}`, title: label.slice(0, 20) },
            })),
          },
        },
      },
      { headers: authHeaders() }
    );
  } catch {
    // Fallback a texto plano si los botones no están disponibles
    await sendMessage(to, `${bodyText}\n\nOpciones: ${buttons.join(' / ')}`);
  }
}

/**
 * Obtiene la URL de descarga de un media enviado por el usuario.
 * @param {string} mediaId
 * @returns {Promise<string>} URL temporal de descarga
 */
async function getMediaUrl(mediaId) {
  const response = await axios.get(`${BASE_URL}/${mediaId}`, {
    headers: authHeaders(),
  });
  return response.data.url;
}

/**
 * Descarga un media y lo retorna como Buffer.
 * @param {string} mediaUrl
 * @returns {Promise<Buffer>}
 */
async function downloadMedia(mediaUrl) {
  const response = await axios.get(mediaUrl, {
    headers: { Authorization: `Bearer ${config.whatsapp.token}` },
    responseType: 'arraybuffer',
  });
  return Buffer.from(response.data);
}

/**
 * Marca un mensaje como "leído" (✓✓ azul).
 * @param {string} messageId
 */
async function markAsRead(messageId) {
  await axios.post(
    `${BASE_URL}/${config.whatsapp.phoneNumberId}/messages`,
    {
      messaging_product: 'whatsapp',
      status: 'read',
      message_id: messageId,
    },
    { headers: authHeaders() }
  ).catch(() => {}); // No crítico
}

module.exports = { sendMessage, sendButtonMessage, getMediaUrl, downloadMedia, markAsRead };
